library/alloc/src/collections/fixed_queue/mod.rs

```diff
@@ -0,0 +1,288 @@
+    /// Creates a FixedQueue from a fixed size slice.
+    fn from(array: &[T; N]) -> Self {
+        FixedQueue::with(array.map(Some), 0, 0, N)
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T: Copy, const N: usize> From<&[T]> for FixedQueue<T, N> {
+    /// Creates a FixedQueue from an unsized slice. Copies a maximum of N
+    /// elements of the slice, and a minimum of the slice length into the
+    /// queue. {[0, 0], 0} - [0, 0]
+    fn from(array: &[T]) -> Self {
+        let mut buf: [Option<T>; N] = [None; N];
+        let length = N.min(array.len());
+        for i in 0..length {
+            buf[i] = Some(array[i]);
+        }
+        FixedQueue::with(buf, 0, array.len() - 1, length)
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T: PartialEq, const N: usize> PartialEq for FixedQueue<T, N> {
+    /// This method tests if a FixedQueue is equal to another FixedQueue.
+    fn eq(&self, other: &FixedQueue<T, N>) -> bool {
+        if other.len != self.len {
+            return false;
+        }
+        (0..N).all(|x| self.buffer[(self.head + x) % N] == other.buffer[(other.head + x) % N])
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T: PartialEq, const N: usize, const M: usize> PartialEq<[T; M]> for FixedQueue<T, N> {
+    /// This method tests if a FixedQueue is equal to a fixed size array.
+    fn eq(&self, other: &[T; M]) -> bool {
+        if M != self.len {
+            return false;
+        }
+        (0..M).all(|x| self.buffer[(self.head + x) % N].as_ref() == Some(&other[x]))
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T, const N: usize> Index<usize> for FixedQueue<T, N> {
+    type Output = Option<T>;
+
+    fn index(&self, index: usize) -> &Self::Output {
+        if index >= N {
+            panic!("Index out of bounds");
+        }
+        &self.buffer[(self.head + index) % N]
+    }
+}
+
+#[unstable(feature = "fixed_queue", issue = "126204")]
+impl<T, const N: usize> Index<Range<usize>> for FixedQueue<T, N>
+where
+    T: Copy + Default,
+{
+    type Output = [T];
+
+    fn index(&self, range: Range<usize>) -> &Self::Output {
+        let start = range.start;
+        let end = range.end;
+
+        // check bounds
+        assert!(start <= end && end <= self.len, "Index out of bounds");
+
+        // create temporary array to store the results
+        let mut temp = Vec::with_capacity(end - start);
+
+        for i in start..end {
+            let idx = (self.head + i) % N;
+            if let Some(value) = self.buffer[idx] {
+                temp.push(value);
+            }
+        }
+
+        // Return a slice from the temporary array
+        // SAFETY: This is safe because temp will live long enough within this function call.
+        let result = unsafe { slice::from_raw_parts(temp.as_ptr(), temp.len()) };
+        mem::forget(temp);
+        result
```

Comment:
This creates a memory leak for performing simple indexing. This should successfully run to completion if you use `fixed_queue[0..N]` on a FixedQueue of 65535 elements a total of 2147483647 times in a program, without increasing memory usage.